tools/exploredb/main.go

```diff
@@ -95,143 +175,166 @@ func showBucketStats(dbNameWithPath string) {
+			statsC <- stat
+		}(bName)
 	}
+	wg.Wait()
+	close(statsC)
 }
 
-func printBucketContents(dbNameWithPath string, rowLimit uint64, bucketName string) {
-	// get the keys within the supplied limit for the given bucket.
-	bucketNameInBytes := []byte(bucketName)
-	keys, sizes := keysOfBucket(dbNameWithPath, bucketNameInBytes, rowLimit)
-
-	// create a new KV Store.
-	dbDirectory := filepath.Dir(dbNameWithPath)
-	db, openErr := kv.NewKVStore(context.Background(), dbDirectory, &kv.Config{})
-	if openErr != nil {
-		log.Fatalf("could not open db, %v", openErr)
+func readStates(ctx context.Context, db *kv.Store, stateC chan<- *modifiedState, keys [][]byte, sizes []uint64) {
+	for rowCount, key := range keys {
+		st, stateErr := db.State(ctx, bytesutil.ToBytes32(key))
+		if stateErr != nil {
+			log.Errorf("could not get state for key : , %v", stateErr)
+		}
+		mst := &modifiedState{
+			state:     st,
+			key:       key,
+			valueSize: sizes[rowCount],
+			rowCount:  uint64(rowCount),
+		}
+		stateC <- mst
 	}
+	close(stateC)
+}
 
-	// dont forget to close it when ejecting out of this function.
-	defer func() {
-		closeErr := db.Close()
-		if closeErr != nil {
-			log.Fatalf("could not close db, %v", closeErr)
+func readStateSummary(ctx context.Context, db *kv.Store, stateSummaryC chan<- *modifiedStateSummary, keys [][]byte, sizes []uint64) {
+	for rowCount, key := range keys {
+		ss, ssErr := db.StateSummary(ctx, bytesutil.ToBytes32(key))
+		if ssErr != nil {
+			log.Errorf("could not get state summary for key : , %v", ssErr)
 		}
-	}()
+		mst := &modifiedStateSummary{
+			slot:      ss.Slot,
+			root:      ss.Root,
+			key:       key,
+			valueSize: sizes[rowCount],
+			rowCount:  uint64(rowCount),
+		}
+		stateSummaryC <- mst
+	}
+	close(stateSummaryC)
+}
 
-	// retrieve every element for keys in the list and call the respective display function.
-	ctx := context.Background()
-	rowCount := uint64(0)
-	for index, key := range keys {
-		switch bucketName {
-		case "state":
-			printState(ctx, db, key, rowCount, sizes[index])
-		case "state-summary":
-			printStateSummary(ctx, db, key, rowCount)
+func printBucketStates(statsC <-chan *bucketStat, doneC chan<- bool) {
+	for stat := range statsC {
+		if stat.noOfRows != 0 {
+			averageValueSize := stat.totalValueSize / stat.noOfRows
+			averageKeySize := stat.totalKeySize / stat.noOfRows
+			fmt.Println("------ ", stat.bucketName, " --------")
+			fmt.Println("NumberOfRows     = ", stat.noOfRows)
+			fmt.Println("TotalBucketSize  = ", humanize.Bytes(stat.totalValueSize+stat.totalKeySize))
+			fmt.Println("KeySize          = ", humanize.Bytes(stat.totalKeySize), "(min = "+humanize.Bytes(stat.minKeySize)+", avg = "+humanize.Bytes(averageKeySize)+", max = "+humanize.Bytes(stat.maxKeySize)+")")
+			fmt.Println("ValueSize        = ", humanize.Bytes(stat.totalValueSize), "(min = "+humanize.Bytes(stat.minValueSize)+", avg = "+humanize.Bytes(averageValueSize)+", max = "+humanize.Bytes(stat.maxValueSize)+")")
```

Comment:
Can we use `log` for these?

Reply:
Same comment for the rest of the tool. Logrus gives a bit more flexibility with the log output formats.

Reply:
Sure.